### ISSUE TITLE
Better waypointing algorithm

### ISSUE THREAD

**Issue by Lyberta** [red-eclipse/base#571]
Current waypoint algorithm works like this:
1. While you are near waypoint, it remembers where it is.
2. If you walk N units far from it, next waypoint is created and linked to the previous one.

The problems:
1. It doesn't automatically check if other waypoints are reachable.
2. It produces wrong connections if there are 2 waypoints on each side of a thin wall.

I was making my own bots of Team Fortress 2 before Valve released their own. Here's the algorithm I've used when a player requests creation of a new waypoint:
1. For each waypoint trace player hull to it to determine if waypoint is visible from the current position.
2. For each visible waypoint trace player hull downwards every N units along the vector between waypoints and record the collision position.
3. If collision occurred more than M units, this means there is a hole in the floor and the player can't walk through it.
4. If no collision happened M units deep, assume that the player can walk from current waypoint to the other one. Make a link.

For current RE realities, we can add another step first:
0. Find all waypoints that are in the proximity of current one.

So, in a nutshell: find all close waypoints, find all visible ones, find all walkable ones, link them.
